lib/AST/DiagnosticConsumer.cpp

```diff
@@ -14,15 +14,148 @@
 //
 //===----------------------------------------------------------------------===//
 
-#define DEBUG_TYPE "swift-basic"
+#define DEBUG_TYPE "swift-ast"
 #include "swift/AST/DiagnosticConsumer.h"
 #include "swift/AST/DiagnosticEngine.h"
+#include "swift/Basic/Defer.h"
+#include "swift/Basic/SourceManager.h"
 #include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/StringSet.h"
 #include "llvm/Support/Debug.h"
 #include "llvm/Support/raw_ostream.h"
 using namespace swift;
 
-DiagnosticConsumer::~DiagnosticConsumer() { }
+DiagnosticConsumer::~DiagnosticConsumer() = default;
+
+llvm::SMLoc DiagnosticConsumer::getRawLoc(SourceLoc loc) {
+  return loc.Value;
+}
+
+FileSpecificDiagnosticConsumer::FileSpecificDiagnosticConsumer(
+    SmallVectorImpl<ConsumerPair> &consumers)
+  : SubConsumers(std::move(consumers)) {
+#ifndef NDEBUG
+  llvm::StringSet<> seenFiles;
+  for (const ConsumerPair &consumerPair : SubConsumers) {
+    if (consumerPair.first.empty())
+      continue;
+    bool isUnique = seenFiles.insert(consumerPair.first).second;
+    assert(isUnique &&
+           "duplicate file names given to FileSpecificDiagnosticConsumer");
+  }
+#endif
+}
+
+void
+FileSpecificDiagnosticConsumer::buildLocationToConsumerMap(SourceManager &SM) {
+  using MapEntry = std::pair<CharSourceRange, unsigned>;
+
+  // Look up each file's source range and add it to the "map" (to be sorted).
+  size_t i = 0;
+  for (const ConsumerPair &pair : SubConsumers) {
+    SWIFT_DEFER { ++i; };
+    if (pair.first.empty())
+      continue;
+
+    Optional<unsigned> bufferID = SM.getIDForBufferIdentifier(pair.first);
+    assert(bufferID.hasValue() && "consumer registered for unknown file");
+    CharSourceRange range = SM.getRangeForBuffer(bufferID.getValue());
+    LocationToConsumerMap.emplace_back(range, i);
+  }
+
+  // Sort the "map" by buffer /end/ location, for use with std::lower_bound
+  // later. (Sorting by start location would produce the same sort, since the
+  // ranges must not be overlapping, but since we need to check end locations
+  // later it's consistent to sort by that here.)
+  std::sort(LocationToConsumerMap.begin(), LocationToConsumerMap.end(),
+            [](const MapEntry &left, const MapEntry &right) -> bool {
+    auto compare = std::less<const char *>();
+    return compare(getRawLoc(left.first.getEnd()).getPointer(),
+                   getRawLoc(right.first.getEnd()).getPointer());
+  });
+
+  // Check that the ranges are non-overlapping.
```

Comment:
What part of the compiler guarantees non-overlap? What would be the harm? Would this code fall over and die, or just possibly misfile a diagnostic? If in the future, someone introduces a bug that produces overlapping ranges, do we want the compiler to crash or to merely whine?

Reply:
Misfiled diagnostics, but yes, I think we still want the compiler to crash in such a case when assertions are enabled.